HPCC-21181 Add GetFileDetails option to WsDfuXRef.DFUXRefUnusedFiles

wangkx opened the pull request:
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-13664, label: h21181 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 71b301729d8330365246bab14e924034548823c0
Estimated completion time is ~0.5 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-13.ca-central-1.compute.internal

wangkx commented:
@afishbeck please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-13.ca-central-1.compute.internal
Sha: 71b301729d8330365246bab14e924034548823c0
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |53 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |25 sec  (00:00:25) |
|test (hthor) |874 |874 |0 |542 sec  (00:09:02) |
|test (thor) |812 |812 |0 |823 sec  (00:13:43) |
|test (roxie) |953 |953 |0 |392 sec  (00:06:32) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  655 sec (00:10:55) |  125 sec (00:02:05) |   23 sec (00:00:23) |   19 sec (00:00:19) | 2090 sec (00:34:50) |   18 sec (00:00:18) | 2939 sec (00:48:59) |

HPCCSmoketest commented:
Process of PR-13664, label: h21181 starts now.
The reason of this test is: Did not build yet.
Commit ID: 71b301729d8330365246bab14e924034548823c0
Estimated completion time is ~0.9 hour(s)
centos 7.7.1908 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.7.1908 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 71b301729d8330365246bab14e924034548823c0
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |55 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |33 sec  (00:00:33) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |874 |874 |0 |629 sec  (00:10:29) |
|test (thor) |812 |812 |0 |1003 sec  (00:16:43) |
|test (roxie) |953 |953 |0 |713 sec  (00:11:53) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   14 sec (00:00:14) |  903 sec (00:15:03) |  122 sec (00:02:02) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2706 sec (00:45:06) |   18 sec (00:00:18) | 3803 sec (01:03:23) |

wangkx committed:
HPCC-21181 Add GetFileDetails option to WsDfuXRef.DFUXRefUnusedFiles

1. The DFULogicalFile is moved from ws_dfu.ecm to ws_dfu_common.ecm
so that it can be shared with ws_dfuXRef.ecm;
2. The 5 CWsDfuEx methods related to getLogicalFilesSorted() are
moved to new namespace WsDFUHelpers so that they can be shared with
ws_dfuXRefService.cpp;
3. other changes.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-13664, label: h21181 starts now.
The reason of this test is: New commit.
Commit ID: 4f7b50224e1b302b36542a928e33319c99f289be
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Process of PR-13664, label: h21181 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 4f7b50224e1b302b36542a928e33319c99f289be
Estimated completion time is ~0.5 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-37.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-37.ca-central-1.compute.internal
Sha: 4f7b50224e1b302b36542a928e33319c99f289be
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |135 |135 |0 |0 |0 |54 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |17 sec  (00:00:17) |
|test (hthor) |874 |874 |0 |367 sec  (00:06:07) |
|test (thor) |812 |812 |0 |737 sec  (00:12:17) |
|test (roxie) |953 |953 |0 |383 sec  (00:06:23) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  636 sec (00:10:36) |  119 sec (00:01:59) |   22 sec (00:00:22) |   20 sec (00:00:20) | 1809 sec (00:30:09) |   19 sec (00:00:19) | 2635 sec (00:43:55) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 4f7b50224e1b302b36542a928e33319c99f289be
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |135 |135 |0 |0 |0 |55 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |874 |874 |0 |628 sec  (00:10:28) |
|test (thor) |812 |812 |0 |1027 sec  (00:17:07) |
|test (roxie) |953 |953 |0 |709 sec  (00:11:49) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  923 sec (00:15:23) |  124 sec (00:02:04) |   23 sec (00:00:23) |   18 sec (00:00:18) | 2721 sec (00:45:21) |   17 sec (00:00:17) | 3835 sec (01:03:55) |

wangkx commented:
@afishbeck this PR has been modified as we discussed. Please review.

afishbeck commented on esp/services/ws_dfu/ws_dfuHelpers.cpp:
```diff
@@ -0,0 +1,198 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
2020

afishbeck reviewed (changes requested):
@wangkx One minor copyright date issue, otherwise looks fine.

HPCCSmoketest commented:
Process of PR-13664, label: h21181 starts now.
The reason of this test is: Forced to re-test.
Commit ID: b965f94500b509be039595fa6efeee7db8841313
Estimated completion time is ~0.5 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-141.ca-central-1.compute.internal

HPCCSmoketest commented:
Process of PR-13664, label: h21181 starts now.
The reason of this test is: New commit.
Commit ID: b965f94500b509be039595fa6efeee7db8841313
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-141.ca-central-1.compute.internal
Sha: b965f94500b509be039595fa6efeee7db8841313
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |135 |135 |0 |0 |0 |54 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |874 |874 |0 |374 sec  (00:06:14) |
|test (thor) |812 |812 |0 |755 sec  (00:12:35) |
|test (roxie) |953 |953 |0 |416 sec  (00:06:56) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  652 sec (00:10:52) |  118 sec (00:01:58) |   22 sec (00:00:22) |   19 sec (00:00:19) | 1874 sec (00:31:14) |   18 sec (00:00:18) | 2712 sec (00:45:12) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: b965f94500b509be039595fa6efeee7db8841313
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |135 |135 |0 |0 |0 |55 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |874 |874 |0 |582 sec  (00:09:42) |
|test (thor) |812 |812 |0 |1029 sec  (00:17:09) |
|test (roxie) |953 |953 |0 |712 sec  (00:11:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  927 sec (00:15:27) |  125 sec (00:02:05) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2656 sec (00:44:16) |   17 sec (00:00:17) | 3774 sec (01:02:54) |

wangkx commented:
@afishbeck the copyright date issue is fixed now.

afishbeck reviewed (approved):
@wangkx looks good.

wangkx commented:
@richardkchapman this PR is ready to be merged.
